src/aead/chacha20_poly1305.rs

```diff
@@ -51,31 +51,35 @@ fn chacha20_poly1305_seal(
     cpu_features: cpu::Features,
 ) -> Tag {
     #[cfg(target_arch = "x86_64")]
-    if has_sse41(cpu_features) {
-        let key_block = combine_key_and_nonce(key, nonce);
-
-        extern "C" {
-            fn chacha20_poly1305_seal_asm(
-                in_out: *mut u8,
-                avx2_and_bmi2_capable: u64,
-                len_in: usize,
-                ad: *const u8,
-                len_ad: usize,
-                keyp: &[u32; 12],
-            ) -> TagReturn;
-        }
-
-        return unsafe {
-            chacha20_poly1305_seal_asm(
-                in_out.as_mut_ptr(),
-                if has_avx2_bmi2(cpu_features) { 1 } else { 0 },
-                in_out.len(),
-                aad.as_ref().as_ptr(),
-                aad.as_ref().len(),
-                &key_block,
-            )
+    {
+        if has_sse41(cpu_features) {
+            let mut key_block = combine_key_and_nonce(key, nonce);
+
+            extern "C" {
+                // This function stores the calculated Tag in keyp[8..12] and returns a reference to it
+                fn GFp_chacha20_poly1305_seal<'ctx>(
+                    in_out: *mut u8,
+                    len_in: usize,
+                    ad: *const u8,
+                    len_ad: usize,
+                    keyp: &'ctx mut [u32; 12],
+                ) -> &'ctx [u8; 16];
```

Comment:
Should we use `__m128i` as the return type? Then it would get returned in xmm0 and we could avoid copying it around.

Reply:
Returning `__m128i` from FFI is an experimental feature.